File: app/lib/domain/streak_calculator.dart
class StreakUpdate {
  final int newCurrentDays;
  final int newLongestDays;
  final int freezesConsumed;
  final bool streakBroken;
  const StreakUpdate({
    required this.newCurrentDays,
    required this.newLongestDays,
    required this.freezesConsumed,
    required this.streakBroken,
  });
}

StreakUpdate evaluateStreakOnActive({
  required int currentStreakDays,
  required int longestStreakDays,
  required int freezesRemaining,
  required DateTime? lastActiveDate,
  required DateTime now,
}) {
  if (lastActiveDate == null) {
    return StreakUpdate(
      newCurrentDays: 1,
      newLongestDays: longestStreakDays > 1 ? longestStreakDays : 1,
      freezesConsumed: 0,
      streakBroken: false,
    );
  }

  final daysSince = _daysBetween(lastActiveDate, now);

  if (daysSince == 0) {
    return StreakUpdate(
      newCurrentDays: currentStreakDays,
      newLongestDays: longestStreakDays,
      freezesConsumed: 0,
      streakBroken: false,
    );
  }

  if (daysSince == 1) {
    final newCurrent = currentStreakDays + 1;
    final newLongest = newCurrent > longestStreakDays ? newCurrent : longestStreakDays;
    return StreakUpdate(
      newCurrentDays: newCurrent,
      newLongestDays: newLongest,
      freezesConsumed: 0,
      streakBroken: false,
    );
  }

  // Missed >= 2 days
  final freezesToUse = daysSince - 1;
  if (freezesRemaining >= freezesToUse) {
    final newCurrent = currentStreakDays + 1;
    final newLongest = newCurrent > longestStreakDays ? newCurrent : longestStreakDays;
    return StreakUpdate(
      newCurrentDays: newCurrent,
      newLongestDays: newLongest,
      freezesConsumed: freezesToUse,
      streakBroken: false,
    );
  }

  // Not enough freezes
  final newCurrent = freezesRemaining + 1;
  return StreakUpdate(
    newCurrentDays: newCurrent,
    newLongestDays: longestStreakDays,
    freezesConsumed: freezesRemaining,
    streakBroken: true,
  );
}

bool shouldAwardFreeze({
  required DateTime? lastFreezeAwardDate,
  required DateTime now,
}) {
  if (lastFreezeAwardDate == null) return false;
  return _daysBetween(lastFreezeAwardDate, now) >= 30;
}

int _daysBetween(DateTime from, DateTime to) {
  final fromDay = DateTime(from.year, from.month, from.day);
  final toDay = DateTime(to.year, to.month, to.day);
  return toDay.difference(fromDay).inDays;
}

File: app/android/app/src/main/java/com/gymcoach/gym_coach/domain/streak/StreakCalculator.kt
package com.gymcoach.gym_coach.domain.streak

import java.time.LocalDate
import java.time.temporal.ChronoUnit

/**
 * Mirrors `lib/domain/streak_calculator.dart`. All logic is pure; the
 * repository owns the persistence. Day boundaries use the device's local
 * timezone (no UTC math), matching Drift's behavior.
 */
object StreakCalculator {

    data class StreakUpdate(
        val newCurrentDays: Int,
        val newLongestDays: Int,
        val freezesConsumed: Int,
        val streakBroken: Boolean
    )

    fun evaluateStreakOnActive(
        currentStreakDays: Int,
        longestStreakDays: Int,
        freezesRemaining: Int,
        lastActiveDate: LocalDate?,
        now: LocalDate
    ): StreakUpdate {
        if (lastActiveDate == null) {
            val newLongest = maxOf(longestStreakDays, 1)
            return StreakUpdate(1, newLongest, 0, false)
        }
        val daysSince = daysBetween(lastActiveDate, now).toInt()
        return when {
            daysSince == 0 -> StreakUpdate(currentStreakDays, longestStreakDays, 0, false)
            daysSince == 1 -> {
                val newCurrent = currentStreakDays + 1
                StreakUpdate(newCurrent, maxOf(longestStreakDays, newCurrent), 0, false)
            }
            daysSince >= 2 -> {
                val needed = daysSince - 1
                if (freezesRemaining >= needed) {
                    val newCurrent = currentStreakDays + 1
                    StreakUpdate(newCurrent, maxOf(longestStreakDays, newCurrent), needed, false)
                } else {
                    val newCurrent = freezesRemaining + 1
                    StreakUpdate(newCurrent, longestStreakDays, freezesRemaining, true)
                }
            }
            daysSince < 0 -> {
                // Future-dated lastActive (clock skew): reset to 1, no break flag.
                StreakUpdate(1, maxOf(longestStreakDays, 1), 0, false)
            }
            else -> StreakUpdate(currentStreakDays, longestStreakDays, 0, false)
        }
    }

    /** Returns true iff ≥30 days have passed since `lastFreezeAwardDate`. */
    fun shouldAwardFreeze(lastFreezeAwardDate: LocalDate?, now: LocalDate): Boolean {
        if (lastFreezeAwardDate == null) return false
        return daysBetween(lastFreezeAwardDate, now) >= 30
    }

    private fun daysBetween(a: LocalDate, b: LocalDate): Long =
        ChronoUnit.DAYS.between(a, b)
}
